ria_remote/remote.py

```diff
@@ -231,135 +227,245 @@ def __init__(self, host):
+        # be more brutal if it doesn't work
+        if exitcode is None:  # timed out
+            # TODO: Theoretically terminate() can raise if not successful. How to deal with that?
+            self.shell.terminate()
+        self.sshmanager.close()
+
+    def _append_end_markers(self, cmd):
+        """Append end markers to remote command"""
+
+        return cmd + ' && echo "{}" || echo "{}"\n'.format(self.REMOTE_CMD_OK, self.REMOTE_CMD_FAIL)
+
+    def _get_download_size_from_key(self, key):
+        """Get the size of an annex object file from it's key
+
+        Note, that this is not necessarily the size of the annexed file, but possibly only a chunk of it.
+
+        Parameter
+        ---------
+        key: str
+          annex key of the file
+
+        Returns
+        -------
+        int
+          size in bytes
+        """
+        # TODO: datalad's AnnexRepo.get_size_from_key() is not correct/not fitting. Incorporate the wisdom there, too.
+        #       We prob. don't want to actually move this method there, since AnnexRepo would be quite an expensive
+        #       import. Startup time for special remote matters.
+        # TODO: this method can be more compact. we don't need particularly elaborated error distinction
+
+        # see: https://git-annex.branchable.com/internals/key_format/
+        key_parts = key.split('--')
+        key_fields = key_parts[0].split('-')
+
+        s = S = C = None
+
+        for field in key_fields[1:]:  # note: first one has to be backend -> ignore
+            if field.startswith('s'):
+                # size of the annexed file content:
+                s = int(field[1:]) if field[1:].isdigit() else None
+            elif field.startswith('S'):
+                # we have a chunk and that's the chunksize:
+                S = int(field[1:]) if field[1:].isdigit() else None
+            elif field.startswith('C'):
+                # we have a chunk, this is it's number:
+                C = int(field[1:]) if field[1:].isdigit() else None
+
+        if s is None:
+            return None
+        elif S is None and C is None:
+            return s
+        elif S and C:
+            if C <= int(s / S):
+                return S
+            else:
+                return s % S
+        else:
+            raise RemoteError("invalid key: {}".format(key))
+
+    def _run(self, cmd, no_output=True, check=False):
+
+        # TODO: we might want to redirect stderr to stdout here (or have additional end marker in stderr)
+        #       otherwise we can't empty stderr to be ready for next command. We also can't read stderr for better error
+        #       messages (RemoteError) without making sure there's something to read in any case (it's blocking!)
+        #       However, if we are sure stderr can only ever happen if we would raise RemoteError anyway, it might be
+        #       okay
+        call = self._append_end_markers(cmd)
+        self.shell.stdin.write(call.encode())
```

Comment:
Here `encode`s `encoding` parameter defaults to "utf-8".  Should this be `encoding=locale.getpreferredencoding(False)` instead?  Same question applies in a couple of other spots, including the `decode` call below.

Reply:
Not sure. We are talking to the remote shell. So, if anything we would need to figure out preferred encoding on the remote end, I guess. I'm not entirely clear, what direction would be affected by what settings. However, in my estimation it doesn't matter, since we don't deal with arbitrary user input or arbitrary file names here. It's either annex/objects paths or "archives/archive.7z" or one of the version files.